libyul/optimiser/Rematerialiser.cpp

```diff
@@ -78,7 +78,12 @@ void Rematerialiser::visit(Expression& _e)
 			auto const& value = *m_value.at(name);
 			size_t refs = m_referenceCounts[name];
 			size_t cost = CodeCost::codeCost(m_dialect, value);
-			if (refs <= 1 || cost == 0 || (refs <= 5 && cost <= 1) || m_varsToAlwaysRematerialize.count(name))
+			if (
+				(refs <= 1 && m_loopDepth == 0) ||
+				cost == 0 ||
+				(refs <= 5 && cost <= 1) ||
```

Comment:
Should this also take loops into account? Also: should we *not* take loops into account once, then simplify expressions, then loop invariant code motion and only after that take loops into account? Or will that oscillate too much?

Reply:
I have the same question as @ekpyron here. Do we make an exception for cheap replacements?

Reply:
Can variables with multiple references ([2,5]) whose rematerialization cost is under `1` be rematerialized in and across for loops? I guess example is

```
// Cost of rematerializating x is 1
let x = 0xff
// First reference
for {} lt(x, 0x100) {}
{
  // Second reference
  let y := add(x, 1)
}
```

Reply:
I wrote a small test case to satisfy the `refs <= 5 && cost <= 1` condition but not the ones before. It looks as follows

```
{
	// Cost of rematerializating x is 1
	let x := 0xff
	// First reference in loop to be rematerialized
	for {} lt(x, 0x100) {}
	{
		// Second reference in loop to be rematerialized
		let y := add(x, 1)
	}
}
// ====
// step: rematerialiser
// ----
// {
//     let x := 0xff
//     for { } lt(0xff, 0x100) { }
//     { let y := add(0xff, 1) }
// }
```

Reply:
And one more for nested loop

```
{
	// Cost of rematerializating x is 1
	let x := 0xff
	// First reference in loop to be rematerialized
	for {} lt(x, 0x100) {}
	{
		// Second reference in loop to be rematerialized
		let y := add(x, 1)
		// Third reference in nested loop to be rematerialized
		for {} lt(x, 0x200) {}
		{
			let z := mul(x, 2)
		}
	}
}
// ====
// step: rematerialiser
// ----
// {
//     let x := 0xff
//     for { } lt(0xff, 0x100) { }
//     {
//         let y := add(0xff, 1)
//         for { } lt(0xff, 0x200) { }
//         { let z := mul(0xff, 2) }
//     }
// }
```

Reply:
And for the case where there are more than 5 refs

```
{
	// Cost of rematerializating x is 1
	let x := 0xff
	// First reference in loop to be rematerialized
	for {} lt(x, 0x100) {}
	{
		// Second reference in loop to be rematerialized
		let y := add(x, 1)
		let z := mul(x, 1)
		let a := div(x, 2)
		let b := mod(x, 3)
		let c := sdiv(x, 4)
	}
}
// ====
// step: rematerialiser
// ----
// {
//     let x := 0xff
//     for { } lt(x, 0x100) { }
//     {
//         let y := add(x, 1)
//         let z := mul(x, 1)
//         let a := div(x, 2)
//         let b := mod(x, 3)
//         let c := sdiv(x, 4)
//     }
// }
```

Reply:
Thinking of it, I'm not sure if the tuning in this PR sufficiently tackles increase in gas cost due to rematerialization. In the worst case, we may end up adding upto 5 rematerialize ops costing `1` (totaling `5`) per loop iteration.